tests/repository/test_upload.py

```diff
@@ -0,0 +1,121 @@
+import logging
+import time
+import pytest
+import json
+import io
+
+from pubtools.pulplib import (
+    Repository,
+    FileRepository,
+    Task,
+    DetachedException,
+    TaskFailedException,
+)
+
+
+def test_upload_detached():
+    """upload_file raises if called on a detached repo"""
+    with pytest.raises(DetachedException):
+        FileRepository(id="some-repo").upload_file("some-file")
+
+
+def test_upload_file(client, requests_mocker, tmpdir, caplog):
+    """test upload a file to a repo in pulp"""
+
+    logging.getLogger().setLevel(logging.INFO)
+    caplog.set_level(logging.INFO)
+
+    repo_id = "repo1"
+    repo = FileRepository(id=repo_id)
+    repo.__dict__["_client"] = client
+
+    client._CHUNK_SIZE = 20
+
+    request_body = {
+        "_href": "/pulp/api/v2/content/uploads/cfb1fed0-752b-439e-aa68-fba68eababa3/",
+        "upload_id": "cfb1fed0-752b-439e-aa68-fba68eababa3",
+    }
+    upload_id = request_body["upload_id"]
+
+    import_report = {
+        "result": {},
+        "error": {},
+        "spawned_tasks": [{"_href": "/pulp/api/v2/tasks/task1/", "task_id": "task1"}],
+    }
+
+    tasks_report = [{"task_id": "task1", "state": "finished"}]
+
+    somefile = tmpdir.join("some-file.txt")
+    somefile.write(b"there is some binary data:\x00\x01\x02")
+
+    requests_mocker.post(
+        "https://pulp.example.com/pulp/api/v2/content/uploads/", json=request_body
+    )
+    requests_mocker.put(
+        "https://pulp.example.com/pulp/api/v2/contents/%s/0" % upload_id, json=[]
+    )
+    requests_mocker.put(
+        "https://pulp.example.com/pulp/api/v2/contents/%s/20" % upload_id, json=[]
+    )
+    requests_mocker.post(
+        "https://pulp.example.com/pulp/api/v2/repositories/%s/actions/import_upload/"
+        % repo_id,
+        json=import_report,
+    )
+    requests_mocker.post(
+        "https://pulp.example.com/pulp/api/v2/tasks/search/", json=tasks_report
+    )
+    requests_mocker.delete(
+        "https://pulp.example.com/pulp/api/v2/content/uploads/%s/" % upload_id, json=[]
+    )
+
+    assert repo.upload_file(str(somefile)).result() == [
+        Task(id="task1", succeeded=True, completed=True)
+    ]
+
+    # sleep for .1 second to let the whole process finish.
+    time.sleep(0.1)
```

Comment:
I think you've got to loop, this can be too flaky.  You might not be able to reproduce a failure locally, but when this is running on a VM or container in CI which could be shared by many other jobs, it's really possible that 0.1 seconds sometimes isn't enough for the other thread to be scheduled.

I usually would do something like: loop up to 1000 iters with 0.01 seconds each, so test is fast when passing and waits up to 10 seconds before failing.

If there are several tests which would need to do that, then it'd be time for a helper function.